src/main/java/org/ice4j/ice/ComponentSocket.java

```diff
@@ -0,0 +1,206 @@
+    implements PropertyChangeListener
+{
+    /**
+     * The {@link Logger} used by the {@link MergingDatagramSocket} class and
+     * its instances for logging output.
+     */
+    private static final java.util.logging.Logger classLogger
+        = java.util.logging.Logger.getLogger
+                (MergingDatagramSocket.class.getName());
+
+    /**
+     * Controls access to {@link #authorizedAddresses}.
+     */
+    private final Object authorizedAddressesSyncRoot = new Object();
+
+    /**
+     * The set of remote addresses, which this socket is allowed to receive
+     * from. These should be the addresses which we have confirmed (e.g. by
+     * having received a STUN message with correct authentication fields).
+     */
+    private Set<SocketAddress> authorizedAddresses
+        = Collections.newSetFromMap(
+                new ConcurrentHashMap<SocketAddress, Boolean>());
+
+    /**
+     * The owning {@link Component}.
+     */
+    private Component component;
+
+    /**
+     * Whether we have invoked {@link #initializeActive}.
+     */
+    private boolean initializedActive = false;
+
+    /**
+     * The {@link Logger} used by {@link MergingDatagramSocket} instances.
+     */
+    private final Logger logger;
+
+    /**
+     * Initializes a new {@link MergingDatagramSocket} instance.
+     * @throws SocketException
+     */
+    ComponentSocket(Component component, Logger levelDelegate)
+        throws SocketException
+    {
+        super(levelDelegate);
+
+        this.component = component;
+        this.logger = new Logger(classLogger, levelDelegate);
+        component.getParentStream().addPairChangeListener(this);
+    }
+
+    /**
+     * {@inheritDoc}
+     * </p>
+     * Verifies that the source of the packet is an authorized remote address.
+     */
+    protected boolean accept(DatagramPacket p)
+    {
+        return authorizedAddresses.contains(p.getSocketAddress());
+    }
+
+    /**
+     * Adds a specific address to the list of authorized remote addresses.
+     * @param address the address to add.
+     */
+    private void addAuthorizedAddress(SocketAddress address)
+    {
+        synchronized (authorizedAddressesSyncRoot)
+        {
+            if (authorizedAddresses.contains(address))
+            {
+                return;
+            }
+
+            logger.info("Adding allowed address: " + address);
+
+            Set<SocketAddress> newSet
+                = Collections.newSetFromMap(
+                        new ConcurrentHashMap<SocketAddress, Boolean>());
+            newSet.addAll(authorizedAddresses);
+            newSet.add(address);
+
+            authorizedAddresses = newSet;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     * </p>
+     * Handles property change events coming from ICE pairs.
+     * @param event
+     */
+    @Override
+    public void propertyChange(PropertyChangeEvent event)
+    {
+        String propertyName = event.getPropertyName();
```

Comment:
minor nit but technically this could lower (right above where it's used) so we don't have to allocate it if we're not gonna use it anyway